### PR TITLE
More flexible `diff` printing when schema changes type of column

### PR DESCRIPTION
**Pull request by macneale4**
This partially addresses an old bug described in the fixed issue. There are certainly more complete solutions, but this seems like the right amount of effort for something pretty edge casey.

Fixes: https://github.com/dolthub/dolt/issues/8133

### CODE REVIEW

File: go/cmd/dolt/commands/diff.go [hunk 1]
```diff
@@ -1534,13 +1535,73 @@ func unionSchemas(s1 sql.Schema, s2 sql.Schema) sql.Schema {
...
+//   - Going from a low precision float to a high precision float, we'll always take the high precision float.
+//   - Going from a low precision integer to a high precision integer, we'll always take the high precision integer.
+//     Currently, we only support this if the signage is the same.
+//   - Going from a DATE, TIME, or DATETIME to a TIMESTAMP, we'll always take the TIMESTAMP.
+//
+// If none of these rules apply, we'll just take the `a` type.
+//
+// Note this is only for printing the diff. This is not robust for other purposes.
+func chooseMostFlexibleType(a, b sql.Type) sql.Type {
+	at := a.Type()
```
Comment (jycor) [inline go/cmd/dolt/commands/diff.go:1563]: There's a `sql.Type.Promote()` function that might be useful here?
In the case of increasing a column type from int32 to int64.

File: go/cmd/dolt/commands/diff.go [hunk 2]
```diff
@@ -1534,13 +1535,73 @@ func unionSchemas(s1 sql.Schema, s2 sql.Schema) sql.Schema {
...
+	bt := b.Type()
+
+	if bt == at {
+		return a
+	}
+
+	// If both are numbers, we'll take the float.
+	if sqltypes.IsIntegral(at) && sqltypes.IsFloat(bt) {
```
Comment (jycor) [inline go/cmd/dolt/commands/diff.go:1571]: I believe you also need a case for decimal types
```sql
tmp/main*> alter table weights modify column weight decimal(20,10);
tmp/main*> select * from weights;
+----+--------------+
| id | weight       |
+----+--------------+
| 1  | 1.1000000000 |
| 2  | 2.1000000000 |
| 3  | 3.1000000000 |
+----+--------------+
3 rows in set (0.00 sec)

tmp/main*> \diff
diff --dolt a/weights b/weights
--- a/weights
+++ b/weights
 CREATE TABLE `weights` (
   `id` int NOT NULL,
-  `weight` int,
+  `weight` decimal(20,10),
   PRIMARY KEY (`id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_bin;
+---+----+--------+
|   | id | weight |
+---+----+--------+
| < | 1  | 1      |
| > | 1  | 1      |
| < | 2  | 2      |
| > | 2  | 2      |
| < | 3  | 3      |
| > | 3  | 3      |
+---+----+--------+
```

**Review by jycor — APPROVED** [review]
Missing case for `DECIMAL(scale, prec)`
Otherwise, LGTM!

**Comment by coffeegoddd** [comment]
@macneale4 __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>683c651</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>683c651</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**